src/main/scala/mesosphere/marathon/api/serialization/ContainerSerializer.scala

```diff
@@ -2,38 +2,61 @@ package mesosphere.marathon.api.serialization
 
 import mesosphere.marathon.Protos
 import mesosphere.marathon.core.externalvolume.ExternalVolumes
-import mesosphere.marathon.state.Container.Docker
 import mesosphere.marathon.state.Container.Docker.PortMapping
 import mesosphere.marathon.state._
-import org.apache.mesos
+import org.apache.mesos.{ Protos => Mesos }
 
 import scala.collection.JavaConverters._
 import scala.collection.immutable.Seq
 
 object ContainerSerializer {
   def toProto(container: Container): Protos.ExtendedContainerInfo = {
     val builder = Protos.ExtendedContainerInfo.newBuilder
-      .setType(container.`type`)
       .addAllVolumes(container.volumes.map(VolumeSerializer.toProto).asJava)
-    container.docker.foreach { d => builder.setDocker(DockerSerializer.toProto(d)) }
+
+    container match {
+      case _: Container.Mesos =>
+        builder.setType(Mesos.ContainerInfo.Type.MESOS)
+      case docker: Container.Docker =>
+        builder.setType(Mesos.ContainerInfo.Type.DOCKER)
+        builder.setDocker(DockerSerializer.toProto(docker))
+    }
+
     builder.build
   }
 
   def fromProto(proto: Protos.ExtendedContainerInfo): Container = {
-    val maybeDocker = if (proto.hasDocker) Some(DockerSerializer.fromProto(proto.getDocker)) else None
-    Container(
-      `type` = proto.getType,
-      volumes = proto.getVolumesList.asScala.map(Volume(_)).to[Seq],
-      docker = maybeDocker
-    )
+    if (proto.hasDocker) {
```

Comment:
I understand that the docker value makes most sense to evaluate here as it is the only value serialized into the proto. Yet, we might as well deserialize a proto that this code did not serialize; the proto.getType is a required field and is not evaluated at all now. Since the proto has no meaninful validation, I think we should check the type here as well and throw an Exception if type and docker do not match.